container/doctabs.go

```diff
@@ -0,0 +1,298 @@
+		}
+		lastX = offset.X
+		lastY = offset.Y
+		r.updateIndicator(false)
+	}
+	r.updateAllTabs()
+	r.updateCreateTab()
+	r.updateTabs()
+	r.updateIndicator(false)
+	return r
+}
+
+// MinSize returns the size that this widget should not shrink below
+func (t *DocTabs) MinSize() fyne.Size {
+	t.ExtendBaseWidget(t)
+	return t.BaseWidget.MinSize()
+}
+
+// Declare conformity with WidgetRenderer interface.
+var _ fyne.WidgetRenderer = (*docTabsRenderer)(nil)
+
+type docTabsRenderer struct {
+	baseTabsRenderer
+	docTabs  *DocTabs
+	scroller *Scroll
+	box      *fyne.Container
+	create   *widget.Button
+}
+
+func (r *docTabsRenderer) Layout(size fyne.Size) {
+	r.updateAllTabs()
+	r.updateCreateTab()
+	r.updateTabs()
+	r.layout(&r.docTabs.baseTabs, size)
+	r.updateIndicator(true)
+}
+
+func (r *docTabsRenderer) MinSize() fyne.Size {
+	return r.minSize(&r.docTabs.baseTabs)
+}
+
+func (r *docTabsRenderer) Objects() []fyne.CanvasObject {
+	objects := r.baseTabsRenderer.Objects()
+	if i, is := r.docTabs.current, r.docTabs.Items; i >= 0 && i < len(is) {
+		objects = append(objects, is[i].Content)
+	}
+	return objects
+}
+
+func (r *docTabsRenderer) Refresh() {
+	// TODO Offset Scroller so current tab is always visible
+
+	r.Layout(r.docTabs.Size())
+
+	r.refresh(&r.docTabs.baseTabs)
+
+	canvas.Refresh(r.docTabs)
+}
+
+func (r *docTabsRenderer) buildAllTabsButton() (all *widget.Button) {
+	all = widget.NewButton("", func() {
+		// Show pop up containing all tabs
+
+		var items []*fyne.MenuItem
+		for i := 0; i < len(r.docTabs.Items); i++ {
+			item := r.docTabs.Items[i]
+			// FIXME MenuItem doesn't support icons (#1752)
+			// FIXME MenuItem can't show if it is the currently selected tab (#1753)
+			items = append(items, fyne.NewMenuItem(item.Text, func() {
+				r.docTabs.Select(item)
+				r.docTabs.popUp = nil
+			}))
+		}
+
+		r.docTabs.showPopUp(all, items)
+	})
+	all.Importance = widget.LowImportance
+	return
+}
+
+func (r *docTabsRenderer) buildCreateTabsButton() *widget.Button {
+	create := widget.NewButton("", func() {
+		if f := r.docTabs.CreateTab; f != nil {
+			if tab := f(); tab != nil {
+				r.docTabs.Append(tab)
+				r.docTabs.SelectIndex(len(r.docTabs.Items) - 1)
+			}
+		}
+	})
+	create.Importance = widget.LowImportance
+	return create
+}
+
+func (r *docTabsRenderer) buildTabButtons(count int) *fyne.Container {
+	buttons := &fyne.Container{}
+
+	var iconPos buttonIconPosition
+	if fyne.CurrentDevice().IsMobile() {
+		cells := count
+		if cells == 0 {
+			cells = 1
+		}
+		buttons.Layout = layout.NewGridLayout(cells)
+		iconPos = buttonIconTop
+	} else if r.docTabs.tabLocation == TabLocationLeading || r.docTabs.tabLocation == TabLocationTrailing {
+		buttons.Layout = layout.NewVBoxLayout()
+		iconPos = buttonIconTop
+	} else {
+		buttons.Layout = layout.NewHBoxLayout()
+		iconPos = buttonIconInline
+	}
+
+	for i := 0; i < count; i++ {
+		item := r.docTabs.Items[i]
+		button, ok := r.buttonCache[item]
+		if !ok {
+			button = &tabButton{
```

Comment:
I know this was inline in AppTabs, but this is quite a bit more complex now, perhaps a button constructor func will help reduce the complexity of this function?

Reply:
Simplified by moving closing code to method